server/api/classyear/classyear.controller.js

```diff
@@ -220,5 +222,20 @@ function generateCode(codeLength){
       var character = (Math.floor(Math.random() * characterOptions.length));
       code = code.concat(characterOptions[character.toString()]);
   }
+  ClassYear.findOne({"dayCodes.code":code})
+    .exec(function(err, classYear){
+      if (err) return handleError(res, err);
+      if(classYear) {
+        generateCode(codeLength);
+      }
+      else{
+        SmallGroup.findOne({"dayCodes.code":code})
+          .exec(function(err, smallgroup){
+            if (err) return handleError(res, err);
+            if(smallgroup){
+              generateCode(codeLength);
```

Comment:
If this function was a correct async one, you should have a `return generateCode(codeLength)`